### ISSUE TITLE
fix(privacy): DP experiments show unrealistic perfect convergence (cosine=1.0)

### ISSUE THREAD

**Issue by reinesaj2** [reinesaj2/federated-ids#83]
## Problem

All 3 DP-enabled experiments (dp=1, noise sigma=1.01) achieve cosine_similarity=1.0 within 2-3 rounds, which contradicts expected noisy behavior from differential privacy.

**Evidence:**

**Experiment 1:** comp_fedavg_alpha0.5_adv0_dp1_pers0_seed42
- Round 1: cos=0.999982
- Rounds 5-20: cos=1.000000 (perfect)

**Experiment 2:** comp_fedavg_alpha0.5_adv0_dp1_pers0_seed43  
- Rounds 1-20: cos=1.000000 (perfect from start)

**Experiment 3:** comp_fedavg_alpha0.5_adv0_dp1_pers0_seed44
- Round 1: cos=0.999982
- Rounds 5-20: cos=1.000000 (perfect)

**Pattern:** 100% of DP experiments show ceiling (3/3)

## Expected Behavior

Differential privacy noise should PREVENT perfect model convergence:
- Expected cosine similarity: 0.92-0.98 (noisy convergence)
- Actual: 1.000000 (identical to non-DP baseline)

This suggests either:
1. DP noise is ineffective (privacy not guaranteed)
2. Measurement artifact in cosine calculation
3. Implementation bug in noise injection

## Impact

**CRITICAL - Invalidates all privacy-utility tradeoff claims**

- Issue 59 (privacy-utility curve) depends on measurable DP effect
- Thesis claims about privacy-preserving FL cannot be supported
- May indicate actual privacy violation if noise not applied

### Hypothesis 1: Noise Scale Too Small
```python
# If noise sigma=1.01 is too small relative to gradient magnitude
gradient_norm = 100.0  # typical
noise_sigma = 1.01     # too small?
SNR = gradient_norm / noise_sigma = 99x  # noise is 1% of signal
```

### Hypothesis 2: Noise Not Applied
```python
# Check if DP flag actually triggers noise injection
if use_dp:
    # Is this code path actually executed?
    add_noise_to_gradients(...)
```

### Hypothesis 3: Cosine Metric Artifact
```python
# Cosine normalizes vectors - might cancel noise effect
# noise adds to both num and denom, canceling out
cos = dot(v1+n1, v2+n2) / (norm(v1+n1) * norm(v2+n2))
```

### Hypothesis 4: Over-normalization
```python
# If gradients normalized before cosine, noise removed
normalized = gradient / norm(gradient)  # removes noise magnitude
```

### Phase 1: Debug DP Implementation
- [ ] Add debug logging to DP noise injection code
- [ ] Verify noise is actually added to gradients (print statements)
- [ ] Check noise scale calculation vs theory
- [ ] Validate privacy budget epsilon calculation

### Phase 2: Measure Noise Effect
- [ ] Compute L2 distance for DP vs non-DP (should differ)
- [ ] Plot gradient distributions with/without DP
- [ ] Check update_norm metric (should be larger with DP)
- [ ] Manual verification: run with sigma=10 (should be obvious)

### Phase 3: Alternative Metrics
- [ ] Compute parameter-wise Euclidean distance (unnormalized)
- [ ] Use KL divergence between model predictions
- [ ] Track training loss variance (DP should increase)

### Phase 4: Reproduce Literature
- [ ] Compare implementation to McMahan et al. 2018 DP-FedAvg
- [ ] Verify clip-and-noise order (clip first, then noise)
- [ ] Check sensitivity calculation (L2 clip threshold)

## Acceptance Criteria

- [ ] Root cause identified and documented
- [ ] Either:
  - DP implementation fixed and re-tested, OR
  - Cosine metric replaced with DP-sensitive alternative
- [ ] DP experiments show measurable difference from baseline
- [ ] Privacy-utility curve (Issue 59) shows expected tradeoff
- [ ] Unit tests validate noise injection
- [ ] Documentation explains privacy guarantees

## Recommended Quick Test

```bash
# Run with extreme noise to verify implementation
python scripts/run_single_experiment.py \
  --aggregation fedavg \
  --dp_noise_scale 10.0 \
  --rounds 5

# If cosine still 1.0, DP is broken
# If cosine < 0.9, implementation works but scale wrong
```

## Priority

**P1 - CRITICAL**

Blocks:
- Issue 59 (privacy-utility curve)
- Thesis privacy preservation claims
- Scientific validity of DP implementation

Must resolve before thesis defense.

## Related

- Issue 59: Implement privacy-utility curve
- Issue 78: Plot quality audit umbrella
- Milestone M2: Privacy & Security
- Literature: McMahan et al. 2018 "Learning Differentially Private Recurrent Language Models"
